core/src/main/java/io/undertow/conduits/DeflatingStreamSinkConduit.java

```diff
@@ -129,21 +132,42 @@ public int write(final ByteBuffer src) throws IOException {
                     return 0;
                 }
             }
-            byte[] data = new byte[src.remaining()];
-            src.get(data);
-            preDeflate(data);
-            deflater.setInput(data);
-            Connectors.updateResponseBytesSent(exchange, 0 - data.length);
+            ByteBuffer arrayBuffer;
+            ByteBufferPool pool = exchange.getConnection().getByteBufferPool().getArrayBackedPool();
+            if (inputBytes <= pool.getBufferSize()) {
+                if (inputBuffer == null) {
+                    inputBuffer = pool.allocate();
```

Comment:
note: This is only deallocated in this block when a larger buffer is required, and when the conduit is closed in `freeBuffer`.
Ideally I think the deflateData finally block should include this:

```java
if (deflater.needsInput()) {
    freeInputBuffer();
}
```

However that results in much worse performance due to https://issues.jboss.org/browse/UNDERTOW-1584